### ISSUE TITLE
CQN notification objects are not freed correctly, causing segmentation faults

### ISSUE THREAD

**Issue by K900** [oracle/odpi#96]
The code responsible for freeing CQN message objects contains a typo that causes occasional segfaults, use-after-frees, memory leaks and other related weirdness:

`dpiSubscr.cpp`, line 279:

```
for (j = 0; j < query->numTables; j++) {
    if (query->tables[i].numRows > 0)
        dpiUtils__freeMemory(query->tables[i].rows);
}
```

The code erroneously uses `i` as a counter for freeing tables, where it should be `j`.

**Comment by K900** [comment]
I've also pushed a fixed branch to my fork of the repo, but the policy here prevents me from submitting a pull request.

**Comment by ddevienne** [comment]
Such a bug with not happen with C++11 and range-for :)
At one point, ODPI should take the plunge on modern C++, for a future version.
Any plans in that direction, to use the 8 years old standard now?

**Comment by anthony-tuininga** [comment]
Good catch! It isn't commonly discovered due to the simple fact that most of the time only a single table is being reported. :-)

**Comment by K900** [comment]
The way we encountered the issue was with multiple queries and a single table, making this an out-of-bounds free.

**Comment by K900** [comment]
Also, any chance of backporting this to the stable branch?

**Comment by anthony-tuininga** [comment]
You mean 3.1.x?

**Comment by K900** [comment]
Yes. I've built a fixed version locally, but it would be nice to have an upstream release.

**Comment by anthony-tuininga** [comment]
Can you verify that the changes I just made match the ones you made yourself?

**Comment by K900** [comment]
The diffs seem identical: [mine](https://github.com/K900/odpi/commit/8417ab27deaaad9265f933cb959f67d5a8602f3a.diff), [yours](https://github.com/oracle/odpi/commit/27e76da0376874918d6237a2ebfd452256e2418f.diff).

**Comment by anthony-tuininga** [comment]
Excellent. Thanks.

**Comment by anthony-tuininga** [comment]
Pushed to the v3.1.x branch as well.

**Comment by K900** [comment]
Thanks! Waiting for the next release for this and cx_Oracle then :)

**Comment by anthony-tuininga** [comment]
Sure. I'll give it a few days before making the patch releases for ODPI-C and cx_Oracle, though!

**Comment by K900** [comment]
I've got a long running process over the weekend that's using CQN heavily (and that was crashing every few hours before the fix). I can report back on Monday if it survives :)

**Comment by anthony-tuininga** [comment]
That would be appreciated. Thanks!

**Comment by K900** [comment]
Good news: the process has survived for this whole time - about 4 days total - with no issues.

**Comment by cjbj** [comment]
@K900 that's great.  We'll schedule new releases soon.

**Comment by anthony-tuininga** [comment]
Released in ODPI-C 3.1.2.
